port advanced watcher to react

alisonelizabeth opened the pull request:
This is an initial PR for the advanced watcher pages port to react. I’m a noob at typescript and react hooks, so appreciate any feedback on implementation - as well as design!

There are some TODOs left:
- Previous code had a license validity check on save; need some help to determine best way to port this over
- Redirect back to `/watches` is not implemented yet on save. I started trying to use the existing `kbnUrl` service, but was running into issues. I didn’t spend too much time on this, as I think the ultimate goal is to move to react router.

I left the angular code in there for now for testing/review purposes. I will remove once this is looking to be in good shape.

**Screenshots**

Create new watch
<img width="1166" alt="Screen Shot 2019-03-29 at 11 01 17 AM" src="https://user-images.githubusercontent.com/5226211/55255379-e5dec300-5230-11e9-85b6-730811e3d260.png">

Validation error modals
<img width="1180" alt="Screen Shot 2019-03-29 at 11 04 31 AM" src="https://user-images.githubusercontent.com/5226211/55255405-f98a2980-5230-11e9-92ef-47e35f2853e7.png">

<img width="1154" alt="Screen Shot 2019-03-29 at 11 04 47 AM" src="https://user-images.githubusercontent.com/5226211/55255426-00b13780-5231-11e9-86b5-0ce6ecd4651b.png">

Edit existing watch
<img width="1136" alt="Screen Shot 2019-03-29 at 2 17 40 PM" src="https://user-images.githubusercontent.com/5226211/55255663-8df48c00-5231-11e9-9dc4-f773e9f2473f.png">

Simulate watch
<img width="1149" alt="Screen Shot 2019-03-29 at 2 45 19 PM" src="https://user-images.githubusercontent.com/5226211/55255588-66052880-5231-11e9-8726-951e5f022235.png">

Simulate results
<img width="1762" alt="Screen Shot 2019-03-29 at 2 17 53 PM" src="https://user-images.githubusercontent.com/5226211/55255556-54238580-5231-11e9-8cd2-dd72b167664b.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7416/)

elasticmachine commented:
Pinging @elastic/es-ui

yaronp68 commented:
@alisonelizabeth One more suggestion is to have an input template inside the editor when overriding input
Here's an example
```
{
  "took" : 63,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 1000,
    "max_score" : null,
    "hits" : [ {
      
    }
    ]
  }
}
```

alisonelizabeth commented:
@yaronp68 thanks for the feedback! one concern i might have is the input override is optional so a user would have to delete the sample template if they didn't want it. i guess i'm still not familiar enough to know the most common use cases. 

another possibility might be to include the sample template you provided on the left-hand side with the description (the user could then just copy it, if desired), but that might take up too much space. thoughts?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7622/)

alisonelizabeth commented:
I pushed a commit with the updated trigger override fields, fyi @yaronp68 

<img width="1048" alt="Screen Shot 2019-04-01 at 2 34 47 PM" src="https://user-images.githubusercontent.com/5226211/55351197-ad86f100-548b-11e9-8449-77993bb16249.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7693/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7702/)

yaronp68 commented:
@alisonelizabeth Looks great. I would propose to add toggle button to override input and in case it is toggled off (default?) hide the editor?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7811/)

alisonelizabeth commented:
@yaronp68 yeah, that's another possibility! i'd like to get some more feedback from others on it i think.

alisonelizabeth commented:
@bmcconaghy i pushed another commit with the changes we discussed this morning: added the license validity check and also attempted to use the `kbnUrl` service. It still isn't working correctly for me, maybe you could take a look when you get around to reviewing the code? I basically just passed down `kbnUrl` and `licenseService` as props from `watch_edit_route`. Not sure if this is the best approach - let me know if you know a better way! Thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7883/)

yaronp68 commented:
@bmcconaghy @alisonelizabeth  One more suggestion, I believe will help the user to improve the conditions:
When returning results of the watch can we copy the `input.payload.hits` to a separate search field  + editor  named `ctx.payload.hits`

bmcconaghy commented:
We probably want to have something similar to this in the simulate results flyour to indicate status:
![image](https://user-images.githubusercontent.com/2479496/55555271-eb199300-56b2-11e9-82ae-2acfffe22c05.png)

bmcconaghy commented:
I get a fatal error if I click over to simulate when the JSON in the edit box is invalid.

alisonelizabeth commented:
@bmcconaghy thanks for the feedback.

I'll look into the bug.

As far as the icon, good point. I noticed @pcsanwald created a common component for the watch statuses. I think I'll be able to reuse that - https://github.com/elastic/kibana/blob/a5f6de16a22590d839261c1a6e3645818a2620f1/x-pack/plugins/watcher/public/sections/watch_detail/components/watch_detail/watch_action_status.tsx.

bmcconaghy commented on x-pack/plugins/watcher/common/types/watch_types.ts:
```diff
@@ -0,0 +1,62 @@
+  upstreamJson: any;
+}
+
+export interface BaseWatch {
```
we should probably just move all the models to TS, but we can do that in another PR.

alisonelizabeth replied:
i can work on that next

bmcconaghy commented on x-pack/plugins/watcher/common/types/watch_types.ts:
```diff
@@ -0,0 +1,62 @@
+  startTime: Date;
+  watchStatus: {
+    state: string;
+    actionStatuses: [
```
I think this means that actionStatuses can only be an array with one member of this shape.  I think what you want here is:
```
Array<{ state: string; lastExecutionReason: string; }>

bmcconaghy commented on x-pack/plugins/watcher/public/models/index.d.ts:
```diff
@@ -9,3 +9,20 @@ declare module 'plugins/watcher/models/watch' {
+
+// TODO: Remove once typescript definitions are in EUI
+declare module '@elastic/eui' {
+  export const EuiCodeEditor: React.SFC<any>;
```
wondering why these are needed.  I didn't need to add any eui types for my stuff (maybe I'm missing something)

alisonelizabeth replied:
I think it's only for certain eui components.

see: https://github.com/elastic/kibana/blob/master/TYPESCRIPT.md#how-to-fix-common-typescript-errors

this may not be the correct place to put it. let me know what you think.

bmcconaghy commented on x-pack/plugins/watcher/public/sections/watch_edit/components/json_watch_edit_component.tsx:
```diff
@@ -0,0 +1,83 @@
+        <JsonWatchEditForm
+          kbnUrl={kbnUrl}
+          licenseService={licenseService}
+          setModal={(options: { message: string }) => setModal(options)}
```
wondering if the modal stuff should just get pushed down to this component since it is the only one using it AFAICT

alisonelizabeth replied:
yeah, good call. i actually started doing that already as i was looking into the bug you found.

bmcconaghy commented on x-pack/plugins/watcher/public/sections/watch_edit/components/json_watch_edit_form.tsx:
```diff
@@ -0,0 +1,218 @@
+  JSON: 'watchJson',
+};
+
+const JSON_WATCH_ERROR_MSGS = {
```
I think I'd rather have these defined as separate constants  with clear names (or even just inlined -- are they used in more than one spot?).  It took me a second to parse what was happening with error messages.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/json_watch_edit_simulate.tsx:
```diff
@@ -0,0 +1,473 @@
+
+function getTableData(watch: BaseWatch) {
+  const actions = watch.watch && watch.watch.actions;
+  return map(actions, (action, actionId) => {
```
good point. i think i may have carried this over from the previous logic :) i think the difference is lodash `map` supports objects, which actions is in this case (if i remember correctly). i'll look into it again though and confirm.

alisonelizabeth commented:
@bmcconaghy I pushed a commit that addresses most of your comments - thanks for the feedback! I also fixed an existing watcher bug, https://github.com/elastic/kibana/issues/18296.

*TODOs:*
- Switch `<EuiHealth>` component to `<EuiIcon>` for displaying watch statuses (if it’s ok with you, I’ll hold off on fixing this until Paul’s PR is merged)
- Look into use of lodash `map` in `json_watch_edit_simulate.tsx`
- Just realized I think i need to move the simulate form state up a level, so if you tab from `Simulate` --> `Edit` --> back to `Simulate`, the state persists.

I’ll keep working through these, but didn’t want to block you from making progress on your end if you wanted to go ahead and merge. I also met with @gchaps this morning and have a few changes to make as far as text goes.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8300/)
